### PR TITLE
[TIR][USMP] Greedy memory planning algorithm

### PR DESCRIPTION
**Pull request by manupa-arm**
This commit implements a greedy memory planning algorithms 
using the proposed USMP design.There are two greedy algorithms 
introduced here which use the size and number of conflicts as the criteria.

* Adds few test cases checks for fan-out and linear structures.
* Added a test case for ResNet sub-structure
* Added a test case for MobileNet sub-structure
*  This includes a slight fix for buffer info
    extraction where non-linear network buffers
    owned by the main function should not show
    sporadic liveness.

### CODE REVIEW

**Comment by manupa-arm** [comment]
a friendly ping @mbs-octoml ! this depends on #8468

File: src/tir/usmp/algo/greedy_by_size.cc [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+      return pool_info;
+    }
+  }
+  ICHECK(false) << "TVM USMP Internal Error: no candidate have been selected!";
+  return PoolInfo();
+}
+
+Map<BufferInfo, PoolAllocation> GreedyBySize(const Array<BufferInfo>& buffer_info_arr) {
+  std::vector<BufferInfo> buffer_info_vec;
+  Map<BufferInfo, PoolAllocation> pool_allocations;
```
Comment (mbs-octoml) [inline src/tir/usmp/algo/greedy_by_size.cc:70]: move down to main loop

File: src/tir/usmp/algo/greedy_by_size.cc [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                } else {
+                  return a->conflicts.size() > b->conflicts.size();
+                }
+              }
+              return a->size_bytes > b->size_bytes;
+            });
+
+  for (const auto& buf_info : buffer_info_vec) {
+    std::unordered_map<PoolInfo, size_t, ObjectPtrHash, ObjectPtrEqual> pool_offset_candidates;
+    for (const auto& pool_info : buf_info->pool_candidates) {
```
Comment (mbs-octoml) [inline src/tir/usmp/algo/greedy_by_size.cc:90]: only consider pools large enough
explain you don't need to track assigned buffers here (there zero constant surprised me) since everything is captured by the conflict buffers.

File: src/tir/usmp/algo/greedy_by_size.cc [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  if (max_address <= pool_size) {
+    return true;
+  }
+  return false;
+}
+
+static PoolInfo SelectPlacementPool(
+    const Array<PoolInfo>& pool_candidates,
+    const std::unordered_map<PoolInfo, size_t, ObjectPtrHash, ObjectPtrEqual>& pool_offsets) {
+  for (const auto& pool_info : pool_candidates) {
```
Comment (mbs-octoml) [inline src/tir/usmp/algo/greedy_by_size.cc:59]: kinda a heuristic too, call out?

File: src/tir/usmp/algo/greedy_by_size.cc [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+      auto conflict_buf_info = Downcast<BufferInfo>(conflict_buf_info_obj);
+      size_t next_offset = 0;
+      if (pool_allocations.count(conflict_buf_info)) {
+        auto pool_allocation = pool_allocations[conflict_buf_info];
+        next_offset = pool_allocation->byte_offset + conflict_buf_info->size_bytes;
+        next_offset = round_up_to_byte_alignment(next_offset, conflict_buf_info->alignment->value);
+        if (IsValidPlacement(pool_allocation->pool_info, next_offset,
+                             buf_info->size_bytes->value)) {
+          if (next_offset > pool_offset_candidates[pool_allocation->pool_info]) {
+            pool_offset_candidates[pool_allocation->pool_info] = next_offset;
```
Comment (mbs-octoml) [inline src/tir/usmp/algo/greedy_by_size.cc:106]: this if in particular confuses me

File: src/tir/usmp/algo/greedy_by_size.cc [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  for (const auto& buf_info : buffer_info_vec) {
+    std::unordered_map<PoolInfo, size_t, ObjectPtrHash, ObjectPtrEqual> pool_offset_candidates;
+    for (const auto& pool_info : buf_info->pool_candidates) {
+      if (IsValidPlacement(pool_info, 0, buf_info->size_bytes->value)) {
+        pool_offset_candidates[pool_info] = 0;
+      }
+    }
+
+    for (const auto& conflict_buf_info_obj : buf_info->conflicts) {
+      auto conflict_buf_info = Downcast<BufferInfo>(conflict_buf_info_obj);
```
Comment (mbs-octoml) [inline src/tir/usmp/algo/greedy_by_size.cc:97]: comments under each if will help here.
basically shift the candidate allocation to avoid already committed assignments for conflicting buffers.

**Review by mbs-octoml — APPROVED** [review]
just nits and my usual plea for more comments :-)

**Comment by manupa-arm** [comment]
Thanks! @mbs-octoml .

I finally got around to address the comments and did some cleanup.

**Comment by manupa-arm** [comment]
@mbaret Could you take look at this PR when you have some time ?

File: tests/python/unittest/test_tir_usmp_algo.py [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+
+    def get_allocate(stmt):
+        if isinstance(stmt, tvm.tir.Allocate):
+            allocates[str(stmt.buffer_var.name)] = stmt
+
+    stmt_functor.post_order_visit(primfunc.body, get_allocate)
+    return allocates
+
+
+def assign_poolinfos_to_allocates_in_primfunc(primfunc, pool_infos):
```
Comment (mbaret) [inline tests/python/unittest/test_tir_usmp_algo.py:54]: should these also be 'private' i.e. _assign?

File: tests/python/unittest/test_tir_usmp_algo.py [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+            conflict_str += f'"{conflict.name_hint}", '
+        conflict_str += "]"
+        print(f'_verify_conflicts("{buffer_info_name}", {conflict_str}, buffer_info_map_names)')
+
+
+@pytest.mark.parametrize(
+    ["algorithm", "fast_memory_size", "slow_memory_size"],
+    [("greedy_by_size", 200704, 1418528), ("greedy_by_conflicts", 200704, 1418528)],
+)
+def test_linear(algorithm, fast_memory_size, slow_memory_size):
```
Comment (mbaret) [inline tests/python/unittest/test_tir_usmp_algo.py:185]: As a general comment on these tests, I'd normally favour a larger number of smaller cases. It's quite hard to parse these complex TIR modules and determine the expected behaviour of the test. Additionally, it's ambiguous what coverage we have other than it works in a particular linear and fan-out case.

File: src/tir/usmp/algo/greedy.cc [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    for (const auto& buffer_info : buffer_info_arr) {
+      buffer_info_vec.push_back(std::move(buffer_info));
+    }
+    std::sort(buffer_info_vec.begin(), buffer_info_vec.end(),
+              [](const BufferInfo& a, const BufferInfo& b) {
+                if (a->size_bytes->value == b->size_bytes->value) {
+                  if (a->conflicts.size() == b->conflicts.size()) {
+                    auto a_name_hash = std::hash<std::string>{}(a->name_hint->data);
+                    auto b_name_hash = std::hash<std::string>{}(b->name_hint->data);
+                    return a_name_hash > b_name_hash;
```
Comment (mbaret) [inline src/tir/usmp/algo/greedy.cc:172]: I'm not sure this edge-case behaviour is explicitly tested.

File: src/tir/usmp/analysis/extract_buffer_info.cc [hunk 9]
```diff
@@ -415,18 +454,30 @@ Map<BufferInfo, tir::Stmt> BufferInfoExtractor::operator()(const PrimFunc& main_
 
   // Traverse the liveness events using a open set to track what
   // is live while updating the conflicts through out the linear traversal
-  std::unordered_set<BufferInfo, ObjectPtrHash, ObjectPtrEqual> open_set;
+  //  std::unordered_set<BufferInfo, ObjectPtrHash, ObjectPtrEqual> open_set;
```
Comment (mbaret) [inline src/tir/usmp/analysis/extract_buffer_info.cc:457]: remove

File: src/tir/usmp/analysis/extract_buffer_info.cc [hunk 10]
```diff
@@ -415,18 +454,30 @@ Map<BufferInfo, tir::Stmt> BufferInfoExtractor::operator()(const PrimFunc& main_
...
-      for (const auto& open_buffer_info : open_set) {
+      for (const auto& kv : open_set) {
+        BufferInfo open_buffer_info = kv.first;
         open_buffer_info->conflicts.push_back(le_event.buffer_info);
         if (le_event.buffer_info != open_buffer_info) {
           le_event.buffer_info->conflicts.push_back(open_buffer_info);
         }
       }
-      open_set.insert(le_event.buffer_info);
+      //      open_set.insert(le_event.buffer_info);
```
Comment (mbaret) [inline src/tir/usmp/analysis/extract_buffer_info.cc:468]: remove

File: src/tir/usmp/analysis/extract_buffer_info.cc [hunk 11]
```diff
@@ -415,18 +454,30 @@ Map<BufferInfo, tir::Stmt> BufferInfoExtractor::operator()(const PrimFunc& main_
...
+        open_set[le_event.buffer_info] += 1;
+      }
     } else {
-      open_set.erase(le_event.buffer_info);
+      if (open_set[le_event.buffer_info] == 1) {
+        open_set.erase(le_event.buffer_info);
+      } else {
+        open_set[le_event.buffer_info] -= 1;
+      }
+      //      open_set.erase(le_event.buffer_info);
```
Comment (mbaret) [inline src/tir/usmp/analysis/extract_buffer_info.cc:480]: remove

File: tests/python/unittest/test_tir_usmp_algo.py [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+    )
+    _verify_conflicts("Conv2dOutput_7", ["sid_8", "PaddedInput_7"], buffer_info_map_names)
+
+    _check_max_workspace_size(buffer_pool_allocations, slow_memory_pool, slow_memory_size)
+    _check_max_workspace_size(buffer_pool_allocations, fast_memory_pool, fast_memory_size)
+
+
+# fmt: off
+@tvm.script.ir_module
+class ResnetStructure:
```
Comment (mbaret) [inline tests/python/unittest/test_tir_usmp_algo.py:227]: If the behaviour we're testing is specifically fan-out, I think this can be done with a simpler case than ResNet.

File: src/tir/usmp/analysis/extract_buffer_info.cc [hunk 13]
```diff
@@ -203,29 +214,40 @@ void BufferInfoExtractor::RecordAllocateNodeInfo(const AllocateNode* op) {
...
-    Optional<Target> tgt = func->GetAttr<Target>(tvm::attr::kTarget);
-    ICHECK(tgt) << "There should not be any PrimFuncs without a target attached by now";
-    auto workspace_alignment =
-        tgt.value()->GetAttr<Integer>("workspace-byte-alignment").value_or(16);
-    auto buffer_info = BufferInfo(GetUniqueBufferName(op->buffer_var->name_hint), size_bytes,
-                                  pool_candidates, workspace_alignment);
-    auto allocate = GetRef<Allocate>(op);
-    allocate_var_to_stmt_map_.Set(op->buffer_var, allocate);
-    buffer_info_map_.Set(buffer_info, allocate);
+  if (size_bytes.defined()) {
```
Comment (mbaret) [inline src/tir/usmp/analysis/extract_buffer_info.cc:217]: I think these changes need rationalizing in the commit message as I can't immediately see how they relate to the greedy algorithm.

File: tests/python/unittest/test_tir_usmp_algo.py [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+    tir_mod = assign_poolinfos_to_allocates_in_irmodule(
+        tir_mod, [fast_memory_pool, slow_memory_pool]
+    )
+    main_func = tir_mod["run_model"]
+    buffer_info_map = tvm.tir.usmp.analysis.extract_buffer_info(main_func, tir_mod)
+
+    fcreate_array_bi = tvm.get_global_func("tir.usmp.CreateArrayBufferInfo")
+    buffer_info_arr = fcreate_array_bi(buffer_info_map)
+    fusmp_algo_greedy_by_size = tvm.get_global_func(f"tir.usmp.algo.{algorithm}")
+    buffer_pool_allocations = fusmp_algo_greedy_by_size(buffer_info_arr)
```
Comment (mbaret) [inline tests/python/unittest/test_tir_usmp_algo.py:206]: The naming here (and fusmp_algo_greedy_by_size) is a bit confusing because 'algorithm' could be by conflict I think.

File: src/tir/usmp/algo/greedy.cc [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+      for (const auto& pool_info : buf_info->pool_candidates) {
+        // Mark pool candidates that satisfy the size constraints.
+        if (IsValidPlacement(pool_info, 0, buf_info->size_bytes->value)) {
+          pool_offset_candidates[pool_info] = 0;
+        }
+      }
+
+      for (const auto& conflict_buf_info_obj : buf_info->conflicts) {
+        auto conflict_buf_info = Downcast<BufferInfo>(conflict_buf_info_obj);
+        size_t next_offset = 0;
```
Comment (mbaret) [inline src/tir/usmp/algo/greedy.cc:124]: I've had historic queries around size_t vs. uint64_t in these situations (where sizeof/indexing isn't really being used). I don't have a strong view either way, but wondering if you've got a particular reason to favour size_t here?

File: src/tir/usmp/algo/greedy.cc [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+      const std::unordered_map<PoolInfo, size_t, ObjectPtrHash, ObjectPtrEqual>& pool_offsets) {
+    // Here the pool candidates are ordered when it is consumed by the algorithm.
+    // This could be from order the user has specified. However, schedulers are
+    // welcome to change the order for performance reasons.
+    for (const auto& pool_info : pool_candidates) {
+      if (pool_offsets.count(pool_info)) {
+        return pool_info;
+      }
+    }
+    ICHECK(false) << "TVM USMP Internal Error: no candidate have been selected!";
```
Comment (mbaret) [inline src/tir/usmp/algo/greedy.cc:102]: When would this check be reached (if ever)? It seems like it may be hit if every pool is full, in which case I think the reported error should be quite explicit as it's something the user could reasonably run into. We should also probably test it as I don't think it would be a 'true assert'.

File: src/tir/usmp/algo/greedy.cc [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tir/analysis/usmp/algo/greedy_by_size.cc
```
[the remark this answers is not in the document]
Comment (manupa-arm) [reply src/tir/usmp/algo/greedy.cc:21]: Ack

File: src/tir/usmp/algo/greedy.cc (hunk 16 above)
Comment (manupa-arm) [reply src/tir/usmp/algo/greedy.cc:102]: Yea, make sense, will change.

File: src/tir/usmp/algo/greedy.cc (hunk 15 above)
Comment (manupa-arm) [reply src/tir/usmp/algo/greedy.cc:124]: I favored perf. portability as I dont expect it to overflow.

File: src/tir/usmp/algo/greedy.cc (hunk 8 above)
Comment (manupa-arm) [reply src/tir/usmp/algo/greedy.cc:172]: Well without this; the tests become unstable because of the presence of buffers of same size and conflicts. What would be your suggestion here for a test case ?

File: src/tir/usmp/analysis/extract_buffer_info.cc (hunk 13 above)
Comment (manupa-arm) [reply src/tir/usmp/analysis/extract_buffer_info.cc:217]: https://github.com/apache/tvm/pull/9214/commits/26fffb0fbfac1558bef905dee1817ba2e5eeabdf.

It is mentioned in "a commit" message. I ll capture this into the main message of the PR.
I appreciate that it is truly not part of the greedy algorithm, however getting too small value after the greedy allocation is what actually made me find the issue. 

Therefore, I hope updating the main PR commit message should help here.

File: src/tir/usmp/analysis/extract_buffer_info.cc (hunk 9 above)
Comment (manupa-arm) [reply src/tir/usmp/analysis/extract_buffer_info.cc:457]: Ack.

File: src/tir/usmp/analysis/extract_buffer_info.cc (hunk 10 above)
Comment (manupa-arm) [reply src/tir/usmp/analysis/extract_buffer_info.cc:468]: Ack. Thanks for spotting this!

File: src/tir/usmp/analysis/extract_buffer_info.cc (hunk 11 above)
Comment (manupa-arm) [reply src/tir/usmp/analysis/extract_buffer_info.cc:480]: Thanks!

File: tests/python/unittest/test_tir_usmp_algo.py (hunk 6 above)
Comment (manupa-arm) [reply tests/python/unittest/test_tir_usmp_algo.py:54]: Ack

File: tests/python/unittest/test_tir_usmp_algo.py (hunk 7 above)
Comment (manupa-arm) [reply tests/python/unittest/test_tir_usmp_algo.py:185]: The unit testing in this level is to make sure the component is functional and ensure the contract of the component. As long as greedy algorithm goes covering a linear and fan-out case covers most aspect conflict graph one could end up.
My personal preference would be to check coverage at the integration level of the USMP using these algorithm rather than using hand crafted TIR snippets at this level. WDYT ?

Alternatively, if you had some smaller cases in mind that you'd like tested here, please do share -- so we could consider adding them here.

File: tests/python/unittest/test_tir_usmp_algo.py (hunk 14 above)
Comment (manupa-arm) [reply tests/python/unittest/test_tir_usmp_algo.py:206]: Thanks! Ack -- will change.

File: tests/python/unittest/test_tir_usmp_algo.py (hunk 12 above)
Comment (manupa-arm) [reply tests/python/unittest/test_tir_usmp_algo.py:227]: It has to be a fan-out with varied intermediary buffers as well and getting a snippet from the ResNet model was reasonably easy and seemed reasonable to me. 
Again, if you have suggestion, I would be happy to consider.

**Review by manupa-arm — COMMENTED** [review]
Thanks @mbaret . See my responses and I ll do the acknowledged changes.

File: tests/python/unittest/test_tir_usmp_algo.py (hunk 12 above)
Comment (mbaret) [reply tests/python/unittest/test_tir_usmp_algo.py:227]: Well ResNet has padded convolutions and heavily fused kernels in it with weird names. This makes it quite hard to read - for me at least - and I don't think that complexity enhances a test of fan-out behaviour. It would seem that to test fan-out all we strictly need is a couple of simple elementwise operations that read from the same tensor. I think this test still has merit (to test a more complex 'real' case), but probably needs naming as such (i.e. test_resnet_subgraph).

File: tests/python/unittest/test_tir_usmp_algo.py (hunk 7 above)
Comment (mbaret) [reply tests/python/unittest/test_tir_usmp_algo.py:185]: I think it becomes a bit harder to argue for testing coverage in integration tests when this algorithm component is 'swappable'. That is, unless, those integration tests will be extended for every new algorithm that's added (which is plausible).

Regarding these specific tests, I think they would be a lot more readable using small synthetic tests rather than network extracts. The linear case has a large amount of complexity that doesn't contribute to the test if topology is the only impactful change.

File: src/tir/usmp/algo/greedy.cc (hunk 8 above)
Comment (mbaret) [reply src/tir/usmp/algo/greedy.cc:172]: I'd suggest a small case which explicitly has buffers of the same size and conflicts and confirms they are prioritized according to name hash. It's not immediately obvious from reading the TIR for test_linear or test_fanout that they contain such cases (not saying they necessarily don't), so I think a standalone test is preferable.

File: tests/python/unittest/test_tir_usmp_algo.py (hunk 7 above)
Comment (manupa-arm) [reply tests/python/unittest/test_tir_usmp_algo.py:185]: > That is, unless, those integration tests will be extended for every new algorithm that's added (which is plausible).

Yes, we will be doing this as that would make the most sense as that is the only sensible way to make sure the number workspace size decided by the algorithm, actually turns out to be something that runs.

> Regarding these specific tests, I think they would be a lot more readable using small synthetic tests rather than network extracts. The linear case has a large amount of complexity that doesn't contribute to the test if topology is the only impactful change.

I think, in terms of unit testing this algorithm, I ll create BufferInfo object based tests rather than TIR ones -- truly thats in the input to the algorithm.

File: src/tir/usmp/algo/greedy.cc (hunk 8 above)
Comment (manupa-arm) [reply src/tir/usmp/algo/greedy.cc:172]: As suggested above, I ll create a BufferInfo based test for this one.

Comment (manupa-arm) [reply src/tir/usmp/algo/greedy.cc:172]: Done

File: tests/python/unittest/test_tir_usmp_algo.py (hunk 12 above)
Comment (manupa-arm) [reply tests/python/unittest/test_tir_usmp_algo.py:411]: Done

**Review by manupa-arm — COMMENTED** [review]
@mbaret . I think I've addresssed you comments by now. PTAL when you have some time.

File: src/tir/usmp/algo/greedy.cc [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+      const std::unordered_map<PoolInfo, size_t, ObjectPtrHash, ObjectPtrEqual>& pool_offsets) {
+    // Here the pool candidates are ordered when it is consumed by the algorithm.
+    // This could be from order the user has specified. However, schedulers are
+    // welcome to change the order for performance reasons.
+    for (const auto& pool_info : buf_info->pool_candidates) {
+      if (pool_offsets.count(pool_info)) {
+        return pool_info;
+      }
+    }
+    CHECK(false) << "TVM USMP Error: no candidate have been selected for " << buf_info;
```
Comment (mbaret) [inline src/tir/usmp/algo/greedy.cc:102]: I think this needs to be more obvious because it'll be directly user-facing. Probably it needs to indicate that TVM has exceeded the pool memory limits and suggest to increase them if you want the compilation to pass. Let's not block the whole patch on this though, happy to take it in a follow-up.

**Comment by manupa-arm** [comment]
Ah the CI decided to run itself again, thus I've changed the error message.

**Comment by mbaret** [comment]
This is now merged, thanks @manupa-arm, @mbs-octoml!
